%YAML 1.2
# SPDX-FileCopyrightText: NVIDIA CORPORATION & AFFILIATES
# Copyright (c) 2022 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0
---
name: sync
components:
- name: left_image_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: left_image_receiver
    min_size: 1
- name: right_image_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: right_image_receiver
    min_size: 1
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 12
    policy: 0
- name: left_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: left_cam_receiver
    min_size: 1
- name: right_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: right_cam_receiver
    min_size: 1
- type: nvidia::isaac_ros::Synchronization
  parameters:
    left_image_input: left_image_receiver
    right_image_input: right_image_receiver
    left_cam_input: left_cam_receiver
    right_cam_input: right_cam_receiver
    left_image_output: left_image_transmitter
    right_image_output: right_image_transmitter
    left_cam_output: left_cam_transmitter
    right_cam_output: right_cam_transmitter
    inputs: [left_image_receiver, right_image_receiver, left_cam_receiver, right_cam_receiver]
    outputs: [left_image_transmitter, right_image_transmitter, left_cam_transmitter, right_cam_transmitter]
- name: left_image_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: left_image_transmitter
    min_size: 1
- name: right_image_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: right_image_transmitter
    min_size: 1
- name: left_cam_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: left_cam_transmitter
    min_size: 1
- name: right_cam_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: right_cam_transmitter
    min_size: 1
---
name: disparity
components:
- name: data_receiver_left
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver_left
    min_size: 1
- name: data_receiver_right
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver_right
    min_size: 1
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter
    min_size: 1
- name: pool
  type: nvidia::gxf::BlockMemoryPool
  parameters:
    storage_type: 1
    block_size: 22118400
    num_blocks: 40
- type: nvidia::isaac_ros::SGMDisparity
  parameters:
    left_image_name: "left_image"
    right_image_name: "right_image"
    output_name: "disparity"
    pool: pool
    left_image_receiver: data_receiver_left
    right_image_receiver: data_receiver_right
    output_transmitter: data_transmitter
    min_disparity: 0
    max_disparity: 64
    backends: "CUDA"
---
name: egress
components:
- name: disparity_in
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: disparity_in
    min_size: 1
- name: left_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: left_cam_receiver
    min_size: 1
- name: right_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: right_cam_receiver
    min_size: 1
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter
    min_size: 1
- type: nvidia::isaac_ros::DisparityCompositor
  parameters:
    left_camera_model_receiver: left_cam_receiver
    right_camera_model_receiver: right_cam_receiver
    disparity_receiver: disparity_in
    output_transmitter: data_transmitter
    min_disparity: 0
    max_disparity: 64
---
name: vault
components:
- name: signal
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: signal
    min_size: 1
- name: vault
  type: nvidia::gxf::Vault
  parameters:
    source: signal
    max_waiting_count: 1
    drop_waiting: false
---
name: connections
components:
- type: nvidia::gxf::Connection
  parameters:
    source: sync/left_image_transmitter
    target: disparity/data_receiver_left
- type: nvidia::gxf::Connection
  parameters:
    source: sync/right_image_transmitter
    target: disparity/data_receiver_right
- type: nvidia::gxf::Connection
  parameters:
    source: sync/left_cam_transmitter
    target: egress/left_cam_receiver
- type: nvidia::gxf::Connection
  parameters:
    source: sync/right_cam_transmitter
    target: egress/right_cam_receiver
- type: nvidia::gxf::Connection
  parameters:
    source: disparity/data_transmitter
    target: egress/disparity_in
- type: nvidia::gxf::Connection
  parameters:
    source: egress/data_transmitter
    target: vault/signal
---
components:
- type: nvidia::gxf::GreedyScheduler
  parameters:
    clock: clock
    stop_on_deadlock: false
- name: clock
  type: nvidia::gxf::RealtimeClock